modules/common/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/common/S3.scala

```diff
@@ -42,6 +42,26 @@ object S3 {
 
     def bucketName: String =
       f.split("://").last.split("/").head
+
+    /**
+     * Find diff of two S3 paths.
+     * Return None if parent path is longer than sub path or
+     * parent path doesn't match with sub path.
+     */
+    def diff(parent: Folder): Option[String] = {
+      val parentSplit = parent.split("/")
+      val subSplit = f.split("/")
+      val zipped = parentSplit.zip(subSplit)
+      if (zipped.length != parentSplit.length || zipped.length == subSplit.length) None
+      else {
+        val matched = zipped.foldLeft(true) {
+          case (false, _) => false
+          case (true, (s1, s2)) => s1 == s2
+        }
+        if (matched) Some(subSplit.drop(zipped.length).mkString("/"))
+        else None
+      }
+    }
```

Comment:
There's nothing wrong with your implementation.  I sketched out an alternative implementation only because I find this kind of thing fun.  See what you think:

```scala
def diff(parent: Folder): Option[String] = {

  def go(parentParts: List[String], subParts: List[String]): Option[String] =
    (parentParts, subParts) match {
      case (p, Nil) =>
        Some(p.reverse.mkString("/"))
      case (pHead :: pTail, sHead :: sTail) if pHead === sHead =>
        go(pTail, sTail)
      case _ =>
        None
    }

  go(parent.split("/").reverse, f.split("/").reverse)
}
```

Reply:
Sorry, I got it wrong in previous comment.  I think I meant this:

```scala
def diff(parent: Folder): Option[String] = {

  def go(parentParts: List[String], subParts: List[String], result: List[String]): Option[String] =
    (parentParts, subParts) match {
      case (p, Nil) =>
        Some(result.mkString("/"))
      case (pHead :: pTail, sHead :: sTail) if pHead === sHead =>
        go(pTail, sTail, pHead :: result)
      case _ =>
        None
    }

  go(parent.split("/").reverse, f.split("/").reverse, Nil)

```

Reply:
I think correct implementation is this one:
```scala
    def diff(parent: Folder): Option[String] = {
      def go(parentParts: List[String], subParts: List[String]): Option[String] =
        (parentParts, subParts) match {
          case (_, Nil) =>
            None
          case (Nil, s) =>
            Some(s.mkString("/"))
          case (pHead :: _, sHead :: _) if pHead != sHead =>
            None
          case (pHead :: pTail, sHead :: sTail) if pHead == sHead =>
            go(pTail, sTail)
        }

      go(parent.split("/").toList, f.split("/").toList)
    }
```